### ISSUE TITLE
Function is crashing with sagittal data

### ISSUE THREAD

**Issue by matthieufaillot** [neuropoly/spinalcordtoolbox#2625]
Hello,
I am running SCT 4.2.1 on a OSX 10.12.6

I am having trouble processing dmri data ( acquired sagittally, poor resolution ; Processing of T2 files from same patients works fine). There seems to be a problem with dmri moco which I don't understand  : 
  > Exception: Couldn't find mat.Z0000T0000Warp.nii.gz in mat_b0groups (contents: ['mat.Z0000T00000GenericAffine.mat']

problem can be observed on multiple subjects, adding the flag -g 10 didn't help.

(I am running script from London Course that I hardly modified. Can't find the answer either in issues on github forum or on slides and videos from London course)
Thanks a lot
Matthieu

~~~
> Register. Loop across Z (note: there is only one Z if orientation is axial
> Z=0/0: 100%|####################################| 1/1 [00:00<00:00,  1.98iter/s]
> 
> -------------------------------------------------------------------------------
>   Estimating motion on DW images...
> -------------------------------------------------------------------------------
> 
> Input parameters:
>   Input file ............dwi_averaged_groups.nii
>   Reference file ........dwi_mean_0.nii.gz
>   Polynomial degree .....2
>   Smoothing kernel ......2
>   Gradient step .........1
>   Metric ................MI
>   Sampling ..............0.2
>   Todo ..................estimate_and_apply
>   Mask  .................
>   Output mat folder .....mat_dwigroups
> 
> Data dimensions:
>   34 x 151 x 12 x 2
> 
> Copy file_target to a temporary file...
> sct_convert -i dwi_mean_0.nii.gz -o target.nii.gz
> WARNING: File target.nii.gz already exists. Will overwrite it.
> 
> Register. Loop across Z (note: there is only one Z if orientation is axial
> Z=0/0: 100%|####################################| 2/2 [00:00<00:00,  2.79iter/s]
> 
> Copy b=0 registration matrices...
> Traceback (most recent call last):
>   File "/Users/lebbe/sct_4.2.1/scripts/sct_dmri_moco.py", line 507, in <module>
>     main()
>   File "/Users/lebbe/sct_4.2.1/scripts/sct_dmri_moco.py", line 277, in main
>     fname_data_moco_tmp = dmri_moco(param)
>   File "/Users/lebbe/sct_4.2.1/scripts/sct_dmri_moco.py", line 463, in dmri_moco
>     mat_final + 'mat.Z0000T' + str(index_b0[it]).zfill(4) + ext_mat)
>   File "/Users/lebbe/sct_4.2.1/scripts/sct_utils.py", line 433, in copy
>     % (os.path.basename(src), folder, contents))
> Exception: Couldn't find mat.Z0000T0000Warp.nii.gz in mat_b0groups (contents: ['mat.Z0000T00000GenericAffine.mat']

`#dwi`
#===========================================================================================
cd dwi
# bring t2 segmentation in dmri space to create mask (no optimization)
sct_maths -i dmri.nii.gz -mean t -o dmri_mean.nii.gz
sct_register_multimodal -i ../t2/t2_seg.nii.gz -d dmri_mean.nii.gz -identity 1 -x nn
# create mask to help moco and for faster processing
sct_create_mask -i dmri_mean.nii.gz -p centerline,t2_seg_reg.nii.gz -size 35mm
# crop data
sct_crop_image -i dmri.nii.gz -m mask_dmri_mean.nii.gz -o dmri_crop.nii.gz
# motion correction
# Tips: if data have very low SNR you can increase the number of successive images that are averaged into group with "-g". Also see: sct_dmri_moco -h
sct_dmri_moco -i dmri_crop.nii.gz -bvec bvec.txt -g 10
# segmentation with propseg
sct_deepseg_sc -i dmri_crop_moco_dwi_mean.nii.gz -c dwi -qc ${PATH_QC}
# Register template to dwi
# Tips: Again, here, we prefer to stick to segmentation-based registration. If there are susceptibility distortions in your EPI, then you might consider adding a third step with bsplinesyn or syn transformation for local adjustment.
sct_register_multimodal -i $SCT_DIR/data/PAM50/template/PAM50_t1.nii.gz -iseg $SCT_DIR/data/PAM50/template/PAM50_cord.nii.gz -d dmri_crop_moco_dwi_mean.nii.gz -dseg dmri_crop_moco_dwi_mean_seg.nii.gz -param step=1,type=seg,algo=centermass:step=2,type=seg,algo=bsplinesyn,metric=MeanSquares,smooth=1,iter=3 -initwarp ../t2/warp_template2anat.nii.gz -initwarpinv ../t2/warp_anat2template.nii.gz -qc "$SCT_BP_QC_FOLDER"
# Rename warping fields for clarity
mv warp_PAM50_t12dmri_crop_moco_dwi_mean.nii.gz warp_template2dmri.nii.gz
mv warp_dmri_crop_moco_dwi_mean2PAM50_t1.nii.gz warp_dmri2template.nii.gz
# Warp template and white matter atlas
sct_warp_template -d dmri_crop_moco_dwi_mean.nii.gz -w warp_template2dmri.nii.gz -qc ${PATH_QC}
# Compute DTI metrics
# Tips: The flag -method "restore" allows you to estimate the tensor with robust fit (see: sct_dmri_compute_dti -h)
sct_dmri_compute_dti -i dmri_crop_moco.nii.gz -bval bvals.txt -bvec bvec.txt
# Compute FA within right and left lateral corticospinal tracts from slices 2 to 14 using weighted average method
sct_extract_metric -i dti_FA.nii.gz -z 2:14 -method wa -l 4,5 -o fa_in_cst.csv
# Bring metric to template space (e.g. for group mapping)
sct_apply_transfo -i dti_FA.nii.gz -d $SCT_DIR/data/PAM50/template/PAM50_t2.nii.gz -w warp_dmri2template.nii.gz
# Go back to root folder
cd .. 

> 
~~~

**Comment by jcohenadad** [comment]
@matthieufaillot I can reproduce the bug. Thank you very much for reporting it!

For the records, here is a way to reproduce the bug using SCT example data:
~~~
sct_download_data -d sct_example_data
cd sct_example_data/dmri/
# change orientation to sagittal (e.g. AIL)
sct_image -i dmri.nii.gz -setorient AIL -o dmri_AIL.nii.gz
# run moco and observe the bug
sct_dmri_moco -i dmri_AIL.nii.gz -bvec bvecs.txt -bval bvals.txt 
[...]
Register. Loop across Z (note: there is only one Z if orientation is axial
Z=0/0: 100%|##################################| 10/10 [00:02<00:00,  4.31iter/s]

Copy b=0 registration matrices...
Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_dmri_moco.py", line 507, in <module>
    main()
  File "/Users/julien/code/sct/scripts/sct_dmri_moco.py", line 277, in main
    fname_data_moco_tmp = dmri_moco(param)
  File "/Users/julien/code/sct/scripts/sct_dmri_moco.py", line 463, in dmri_moco
    mat_final + 'mat.Z0000T' + str(index_b0[it]).zfill(4) + ext_mat)
  File "/Users/julien/code/sct/scripts/sct_utils.py", line 433, in copy
    % (os.path.basename(src), folder, contents))
Exception: Couldn't find mat.Z0000T0000Warp.nii.gz in mat_b0groups (contents: ['mat.Z0000T00020GenericAffine.mat', 'mat.Z0000T00030GenericAffine.mat', 'mat.Z0000T00000GenericAffine.mat', 'mat.Z0000T00010GenericAffine.mat', 'mat.Z0000T00040GenericAffine.mat'])
~~~

**Comment by matthieufaillot** [comment]
@jcohenadad  thank you !
I reoriented the dwi images to -LPI

**Comment by jcohenadad** [comment]
@matthieufaillot while the function might work (i.e. not crash) if you reorient your data in LPI, it will not produce the expected results, because when acquired sagittally, motion occurs independently across slices, and is non-rigid within slice, therefore I would not do it (i.e., skip moco) and just wait for the bug fix.

**Comment by matthieufaillot** [comment]
The results of moco were indeed a bit strange but I thought this was due to
the low resolution in the axial plane Thank you
Matthieu


Le lun. 9 mars 2020 à 19:27, Julien Cohen-Adad <notifications@github.com> a
écrit :

**Comment by jcohenadad** [comment]
Digging into `sct_dmri_moco`, it is explicitly mentioned that sagittal scans are not supported: https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_dmri_moco.py#L82

However, it should not be too hard to support that feature. I am working on a fix in #2634

**Comment by jcohenadad** [comment]
Note: the bug also happens with `sct_fmri_moco`
